stone/target/python_rsrc/stone_serializers.py

```diff
@@ -24,11 +24,471 @@
 except (SystemError, ValueError):
     # Catch errors raised when importing a relative module when not in a package.
     # This makes testing this file directly (outside of a package) easier.
-    import stone_validators as bv  # type: ignore
+    import stone_validators as bv # type: ignore
+
+#-------------------------------------------------------------------------
+class StoneEncoderInterface(object):
+    """
+    Interface defining a stone object encoder.
+    """
+
+    def encode(self, validator, value):
+        """
+        Validate ``value`` using ``validator and return the encoding.
+
+        Args:
+            validator: the stone validator used to validate ``value``
+            value: the object to encode
+
+        Returns:
+            This is implementation-defined. The convention is to return an
+            encoded object where ``out_file`` is ``None``, and return
+            ``None`` if ``out_file`` is provided (and supported)
+
+        Raises:
+            stone_validators.ValidationError: Raised if ``value`` (or one
+                of its sub-values).
+        """
+        raise NotImplementedError
+
+    def encode_to_file(self, validator, value, out_file):
```

Comment:
> I can imagine a future where we do the encoding in a streaming fashion which would be cool, but this isn't that.

It's true that implementation doesn't exist yet, but without this function, it's not obvious how that would work.

Serialization tends to risk being expensive. The quick-and-dirty (but interface-compatible) stand-in (for now) is to create the intermediate object hierarchy, then serialize it. But this interface allows a different (direct-to-buffer) implementation in the future while maintaining a (somewhat) sensible interface today?

(At least that was the goal. 😊)